Added support for nested component structure

ijlee2 opened the pull request:
## Description

This PR addresses issue #5. I added support for changing a **classic** component structure to **nested**, all the while preserving the ability to convert to **flat**.

```bash
# Classic to flat
npx github:ember-codemods/ember-component-template-colocation-migrator

# Classic to flat (alternative)
npx github:ember-codemods/ember-component-template-colocation-migrator -fs

# Classic to nested
npx github:ember-codemods/ember-component-template-colocation-migrator -ns
```

I also refactored the `Migrator` class so that we can better maintain it in the future.

## Notes

I recommend reviewing changes one commit at a time.

I apologize in advance for having many commits in one PR. I found the `Migrator` class difficult to understand so I wanted to refactor it before writing code. I ensured that each commit is a small, meaningful change.

If you'd like, I can submit 2 PRs. One for refactor (commits 1-12) and the other for feature (commits 13-17).


## Testing

In addition to adding 2 test cases (a "typical" nested component and a template-only component) for `flat`, I added tests for `nested` (using the same input as `flat`).

I also ran my codemod on 2 production apps. Both are Octane apps that had a mix of classic and pod components. I manually moved pod components to follow the classic structure, then ran the codemod so that all components follow the flat or nested structure. Lastly, I checked that the components are in the right place and ran our test suite. ✅

Note, my production apps didn't have examples of `layoutName` or `{{partial}}`. They did have examples of nested components and template-only components.

ijlee2 committed:
Renamed the folder to classic-to-flat

ijlee2 committed:
Added test examples of a "typical" nested component and a template-only component

ijlee2 committed:
Stored projectRoot as a class field

ijlee2 committed:
Refactored migrator by creating findClassicComponentTemplates and removeEmptyClassicComponentDirectories methods

ijlee2 committed:
Refactored migrator by creating findClassicComponentClasses method

ijlee2 committed:
Shortened variable names for readability

ijlee2 committed:
Refactored migrator by creating skipTemplatesUsedAsLayoutName method

ijlee2 committed:
Refactored migrator by creating findTemplates method

ijlee2 committed:
Refactored migrator by creating skipTemplatesUsedAsPartial method

ijlee2 committed:
Refactored migrator by creating changeComponentStructureToFlat method

ijlee2 committed:
Refactored migrator by adding a better separation of concerns between removeEmptyClassicComponentDirectories and execute

ijlee2 committed:
Moved logging to migrator class methods

ijlee2 committed:
Allowed passing newComponentStructure to Migrator class

ijlee2 committed:
Reorganized test fixtures in preparation for testing nested component structure

ijlee2 committed:
Added test example for migrating to nested component structure

ijlee2 commented on test/fixtures/classic-to-flat/input.js:
```diff
@@ -1,38 +1,78 @@
+      // A standalone component
       'top-level-component.js': '// top-level-component.js',
+
+      // A nested component
```
Lines 9-16 and 45-52 represent a "typical" nested component and are new.

ijlee2 commented on test/fixtures/classic-to-flat/input.js:
```diff
@@ -1,38 +1,78 @@
+        // A standalone component
         'top-level-component.hbs': '{{!-- top-level-component.hbs --}}',
+
+        // A template-only component
```
Lines 42-43 represent a template-only component and are new.

ijlee2 commented on test/fixtures/classic-to-flat/input.js:
```diff
@@ -1,38 +1,78 @@
 module.exports = {
   app: {
     'app.js': '// app',
+
```
I added 2 examples. The remaining changes---whitespaces, comments, and uses of `join()`---were to improve readability.

ijlee2 commented on lib/migrator.js:
```diff
@@ -58,6 +58,35 @@ module.exports = class Migrator {
+        // Extract '/app/templates/components/nested1/nested-component.hbs'
+        let filePathFromApp = templateFilePath.slice(this.projectRoot.length);
+
+        if (/\/\-[\w\-]+\.hbs/.test(filePathFromApp)) {
```
I'm not sure what lines 73-75 were for. This check didn't appear in `skipTemplatesUsedAsLayoutName` and wasn't evaluated to be `true` in tests.

rwjblue replied:
When Ember sees `{{partial "foo"}}` it will look in to locations:

* `app/templates/foo.hbs`
* `app/templates/-foo.hbs`

I think this is attempting to identify when `-foo.hbs` format is used, and still match `{{partial "foo"}}`?

ijlee2 commented on lib/migrator.js:
```diff
@@ -33,6 +33,8 @@ module.exports = class Migrator {
   }
 
   skipTemplatesUsedAsLayoutName(templateFilePaths) {
+    console.info(`\nChecking if any component templates are used as templates of other components using \`layoutName\``);
```
Adding a newline at the beginning helped with making log messages more readable. I think we could spend more time writing logs that are easy to read and understand for the end user.

ijlee2 commented on test/migrator-test.js:
```diff
@@ -10,50 +10,51 @@ var Migrator = require('../lib/migrator');
-  var fixturesPath = path.resolve(__dirname, "fixtures");
+  describe('newComponentStructure = flat', function() {
+    var tmpPath = "tmp/process-files";
+    var fixturesPath = path.resolve(__dirname, "fixtures/classic-to-flat");
```
I updated the directory name to `fixtures/classic-to-flat`. Everything else is a change in indentation.

ijlee2 committed:
Created changeComponentStructureToNested method

ijlee2 committed:
Allowed passing -ns flag to migrate to nested component structure

ijlee2 commented on bin/ember-component-template-colocation-migrator:
```diff
@@ -14,8 +14,19 @@ let options = {
 
+const args = process.argv.slice(2);
+
+// Allow passing the flag, -fs (flat) or -ns (nested), to specify component structure
```
I gave the names `-fs` and `-ns` to match Ember CLI's aliases. I'm not strongly tied to these names, but think they are nice because they are short to type.

rwjblue replied:
👍

rwjblue commented on bin/ember-component-template-colocation-migrator:
```diff
@@ -14,10 +14,26 @@ let options = {
 let projectRoot = parsed['project-root'] || process.cwd();
 
-let migrator = new Migrator({ projectRoot });
+const args = process.argv.slice(2);
```
Let's add `yargs` and use it to parse the command line options.

rwjblue commented on bin/ember-component-template-colocation-migrator:
```diff
@@ -14,10 +14,26 @@ let options = {
+
+let migrator = new Migrator({
+  projectRoot,
+  newComponentStructure
```
Can we call this `structure` (as in this is the structure to target)?

rwjblue commented on bin/ember-component-template-colocation-migrator:
```diff
@@ -14,10 +14,26 @@ let options = {
 }).catch(function(error) {
   console.error(error.stack);
-});
+});
```
Probably should use a newline before the end of file...

rwjblue commented on lib/migrator.js:
```diff
@@ -6,53 +6,158 @@ const { moveFile, removeDirs } = require('./utils/file')
+
+    if (componentsWithLayoutName.length) {
+      componentsWithLayoutName.sort().forEach(component => {
+        console.info(`❌ Did not move '${component}' due to usage as "layoutName" in a component`);
```
We should probably tell them _which_ component

ijlee2 replied:
Sorry, can you elaborate on what you'd like the log message to say?

I only renamed the value for `forEach` from `template` to `component`:
```javascript
// Before
templatesWithLayoutName.sort().forEach(template => {
  console.info(`❌ Did not move '${template}' due to usage as "layoutName" in a component`);
});

// After
componentsWithLayoutName.sort().forEach(component => {
  console.info(`❌ Did not move '${component}' due to usage as "layoutName" in a component`);
});
```

A test run shows that the log message currently looks like:
```
❌ Did not move 'components/layout-name/layout-name-template' due to usage as "layoutName" in a component
```

rwjblue replied:
Ya, I think its fine (didn't notice that it was only a small tweak) for now. I was thinking that saying the template **and** the component that we found using it for `layoutName` would be helpful.

rwjblue commented on lib/migrator.js:
```diff
@@ -6,53 +6,158 @@ const { moveFile, removeDirs } = require('./utils/file')
+
+    if (componentsWithPartial.length) {
+      componentsWithPartial.sort().forEach(component => {
+        console.info(`❌ Did not move '${component}' due to usage as a "partial"`);
```
Should probably say where the `{{partial` invocation was found

ijlee2 replied:
Can you elaborate on what you'd like the log message to say?

A test run shows that the log message currently looks like:
```
❌ Did not move 'components/partials/partials-template' due to usage as a "partial"
```

rwjblue replied:
Basically to say the template name (which the message does now) and the name of hte template that _uses_ it as a partial.

> ❌ Did not move 'components/partials/partials-template' due to usage as a "partial" in 'templates/index.hbs'

rwjblue commented on lib/migrator.js:
```diff
@@ -6,53 +6,158 @@ const { moveFile, removeDirs } = require('./utils/file')
+      moveFile(templateFilePath, newTemplateFilePath);
+
+      // Build '[APP_PATH]/app/components/nested1/nested-component/index.js'
+      const classFilePath = path.join(this.projectRoot, 'app/components', `${targetPath}.js`);
```
Probably should consider `.ts` or `.js`

ijlee2 replied:
I'll look into this. Thanks for the suggestion!

rwjblue commented on lib/migrator.js:
```diff
@@ -6,53 +6,158 @@ const { moveFile, removeDirs } = require('./utils/file')
+    templateFilePaths = this.skipTemplatesUsedAsLayoutName(templateFilePaths);
+    templateFilePaths = this.skipTemplatesUsedAsPartial(templateFilePaths);
+
+    if (this.newComponentStructure === 'flat') {
+      this.changeComponentStructureToFlat(templateFilePaths);
+
+    } else if (this.newComponentStructure === 'nested') {
+      this.changeComponentStructureToNested(templateFilePaths);
+
+    }
```
FWIW, I'd like to make another structure (not sure what a good name for it would be) to use flat when there is only a template or class file and nested when there are both (e.g. to avoid `app/components/foo.js` and `app/components/foo.hbs` living side by side). 

This is definitely out of scope for your PR though, just thinking out loud... 🤔

rwjblue commented on test/fixtures/classic-to-nested/example-1/input.js:
```diff
@@ -0,0 +1,78 @@
+module.exports = {
```
I'm curious, can we get away with a single input that is shared between `classic-to-flat` and `classic-to-nested` then only duplicate that the outputs are different?

ijlee2 replied:
That'd be nice. I can try, but think we can treat this as a low priority.

The input and output files _seemed_ to need to be placed in a particular manner due to the library `fixturify`. I had to create a subfolder in `classic-to-flat` and `classic-to-nested` (which I called `example-1`) and place the files in that subfolder.

rwjblue replied:
> That'd be nice. I can try, but think we can treat this as a low priority.

👍 seems good, would you mind making an issue in the repo for this future cleanup? Just hoping to make it easier for folks to pick it up...

ijlee2 replied:
Yep, will do!

rwjblue reviewed (commented):
This is looking really good!! Only a couple of very minor comments left inline...

ijlee2 commented:
Thanks, @rwjblue! I'll go over your feedback and work on making changes as needed.

ijlee2 committed:
Installed yargs to parse command line options

ijlee2 committed:
Renamed the class field newComponentStructure to structure

ijlee2 committed:
Added a newline at the end of the file

ijlee2 committed:
Documented what the regular expression /\/\-[\w\-]+\.hbs/ does

ijlee2 committed:
Renamed the fixture folder from example-1 to example-js

ijlee2 committed:
Added support for moving component classes written in TypeScript

ijlee2 commented:
@rwjblue When you have time, can you take a look at the changes since last time? Thanks again for your help!

Implemented:
- Used `yargs` to look for `-fs` or `-ns`
- Renamed `newComponentStructure` to `structure`
- Tested a partial template whose name begins with `-`
- Allowed moving component classes written in TypeScript

Did not implement:
- Changed log messages (waiting for your clarification)
- Have shared test input files

rwjblue commented:
Thank you!!!
